cryptofeed/exchanges/binance.py

```diff
@@ -261,26 +272,107 @@ async def _snapshot(self, pair: str) -> None:
+        resp = await self._fetch_snapshot(pair)
 
         std_pair = self.exchange_symbol_to_std_symbol(pair)
         self.last_update_id[std_pair] = resp['lastUpdateId']
         self._l2_book[std_pair] = OrderBook(self.id, std_pair, max_depth=self.max_depth, bids={Decimal(u[0]): Decimal(u[1]) for u in resp['bids']}, asks={Decimal(u[0]): Decimal(u[1]) for u in resp['asks']})
         self._l2_book[std_pair].sequence_number = resp['lastUpdateId']
         self._l2_book[std_pair].raw = resp
+        
+        # Reset messages since snapshot and delete buffer
+        self._msgs_since_snapshot[std_pair] = 0
+        self._fetch_snapshop_buffer.pop(std_pair, None)
+
+    async def _refresh_snapshot(self, pair: str) -> None:
+        resp = await self._fetch_snapshot(pair)
 
-    async def _handle_book_msg(self, msg: dict, pair: str, timestamp: float):
-        """
-        Processes 'depthUpdate' update book msg. This method should only be called if pair's l2_book exists.
-        """
         std_pair = self.exchange_symbol_to_std_symbol(pair)
-        skip_update, current_match = self._check_update_id(std_pair, msg)
-        if current_match:
-            # Current msg.final_update_id == self.last_update_id[pair] which is the snapshot's update id
-            return await self.book_callback(L2_BOOK, self._l2_book[std_pair], timestamp, raw=msg, sequence_number=self.last_update_id[std_pair], timestamp=self.timestamp_normalize(msg['E']))
+        self.last_update_id[std_pair] = resp['lastUpdateId']
 
-        if skip_update:
+        delta = {BID: [], ASK: []}
+        for key, side in [('bids', BID) , ('asks', ASK)]:
+            for price, amount in resp[key]:
+                price = Decimal(price)
+                amount = Decimal(amount)
+                if price in self._l2_book[std_pair].book[side]:
+                    if amount != self._l2_book[std_pair].book[side][price]:
+                        # Since we haven't looked at all the updates, changes in price are allowed here and will be published as deltas
+                        self._l2_book[std_pair].book[side][price] = amount
+                        delta[side].append((price, amount))
+                else:
+                    # Price not in order book: add it there
+                    self._l2_book[std_pair].book[side][price] = amount
+                    delta[side].append((price, amount))
+        self._l2_book[std_pair].sequence_number = resp['lastUpdateId']
+        
+        # Reset messages since snapshot and delete buffer
+        self._msgs_since_snapshot[std_pair] = 0
+        self._fetch_snapshop_buffer.pop(std_pair, None)
+
+        ts = time.time()
+        await self.book_callback(L2_BOOK, self._l2_book[std_pair], ts, timestamp=ts, delta=delta, sequence_number=self.last_update_id[std_pair])
+
+
+    async def _refresh_snapshot_with_buffer(self, pair: str) -> None:
+        resp = await self._fetch_snapshot(pair)
```

Comment:
`ts = time.time()` here for receipt_timestamp (non-named) into `book_callback` below because might be waiting a long time in sleeps during `while resp['lastUpdateId'] > self.last_update_id[std_pair]:`

Reply:
done